### PR TITLE
Add support for (D)TLS 1.2 client hello defragmentation

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Add support for (D)TLS 1.2 client hello defragmentation. Note that there was already some support (but not comprehensive support) for TLS 1.2 ClientHello defragmentation.

~Depends on: https://github.com/Mbed-TLS/mbedtls/pull/10575~ merged

Fixes https://github.com/Mbed-TLS/mbedtls/issues/7549

## PR checklist
- [x] **changelog** provided
- [x] **development PR** here
- [x] **TF-PSA-Crypto PR** not required because: TLS only changes 
- [x] **framework PR** not required
- [x] **3.6 PR** provided #10623
- **tests**  provided

### CODE REVIEW

**Comment by ronald-cron-arm** [comment]
One job on freebsd does not want to start but I think we can consider the CI as green. This is ready for review.

File: programs/ssl/ssl_server2.c [hunk 1]
```diff
@@ -4125,7 +4125,51 @@ int main(int argc, char *argv[])
...
+     * notification. This avoids reconnecting to the same client when we
+     * reset and later receive its close-notification response during
+     * step 3 (waiting for a client to connect).
+     *
+     * Stop waiting for the response if the connection has already ended.
+     */
+#if defined(MBEDTLS_SSL_PROTO_DTLS) && defined(MBEDTLS_HAVE_TIME)
+    if (opt.transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM) {
+        mbedtls_ms_time_t start = mbedtls_ms_time();
+        mbedtls_ssl_conf_read_timeout(&conf, 200);
```
Comment (mpg) [inline programs/ssl/ssl_server2.c]: Note: our documentation [says](https://github.com/Mbed-TLS/mbedtls/blob/75eec4b4772def3321f4cfddf6a5872fb9fd826c/include/mbedtls/ssl.h#L1888) we should not modify a conf object while it's in use. I think it works here, but it's not showing good practice if someone is using this as an example.

If we're doing this we should at least leave a comment.

File: programs/ssl/ssl_server2.c [hunk 2]
```diff
@@ -4041,6 +4045,7 @@ int main(int argc, char *argv[])
         if (opt.serialize == 1) {
             /* nothing to do here, done by context_save() already */
             mbedtls_printf("  . Context has been reset... ok\n");
+            fflush(stdout);
```
Comment (mpg) [inline programs/ssl/ssl_server2.c:4048]: Nit: I don't think this is necessary when what we print ends with a `\n`.

File: library/ssl_tls12_server.c [hunk 3]
```diff
@@ -938,6 +938,9 @@ static int ssl_parse_client_hello(mbedtls_ssl_context *ssl)
         memcpy(&ssl->cur_out_ctr[2], ssl->in_ctr + 2,
                sizeof(ssl->cur_out_ctr) - 2);
 
+    /* Check for record replay and then update the window. This replicates what
+     * is done in `ssl_get_next_record()` when the record is not fetch through
```
Comment (mpg) [inline library/ssl_tls12_server.c]: Typo: fetchED

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -9964,6 +9964,42 @@ run_test    "DTLS reassembly: fragmentation, nbio, renego (gnutls server)" \
...
+            "$G_NEXT_CLI -u --mtu 200 --insecure 127.0.0.1" \
+            0 \
+            -s "found fragmented DTLS handshake message" \
+            -S "error"
+
+requires_gnutls
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+run_test    "DTLS reassembly: more fragmentation (gnutls client)" \
+            "$P_SRV debug_level=2 dtls=1" \
+            "$G_NEXT_CLI -u --mtu 128 --insecure 127.0.0.1" \
```
Comment (mpg) [inline tests/ssl-opt.sh]: From context, I'm assuming this makes the ClientHello fragmented, but otherwise I'm not sure it would be clear to me.

Related: what's the lowest possible value we can pass to `--mtu` here? The lowest the value, the more obvious it is that the ClientHello is fragmented.

File: tests/ssl-opt.sh [hunk 5]
```diff
@@ -9996,6 +10032,30 @@ run_test    "DTLS reassembly: fragmentation, nbio (openssl server)" \
...
+            "$P_SRV debug_level=2 dtls=1" \
+            "$O_NEXT_CLI -dtls -mtu 2048" \
+            0 \
+            -S "found fragmented DTLS handshake message" \
+            -S "error"
+
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+run_test    "DTLS reassembly: some fragmentation (openssl client)" \
+            "$P_SRV debug_level=2 dtls=1" \
+            "$O_NEXT_CLI -dtls -mtu 256" \
```
Comment (mpg) [inline tests/ssl-opt.sh]: Is this enough for the ClientHello to be fragmented? How do we know except by observing this test with wireshark?

File: tests/ssl-opt.sh [hunk 6]
```diff
@@ -11909,6 +12002,39 @@ run_test    "DTLS proxy: 3d, gnutls server, fragmentation, nbio" \
             -s "Extra-header:" \
             -c "Extra-header:"
 
+requires_gnutls
+client_needs_more_time 6
+not_with_valgrind # risk of non-mbedtls peer timing out
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+run_test    "DTLS proxy: 3d, gnutls client" \
+            -p "$P_PXY drop=5 delay=5 duplicate=5" \
```
Comment (mpg) [inline tests/ssl-opt.sh:12241]: Question: why do the test with openssl client have ` protect_hvr=1` but not those with GnuTLS client?

File: programs/ssl/ssl_server2.c [hunk 7]
```diff
@@ -4118,14 +4125,59 @@ int main(int argc, char *argv[])
...
+            if ((ret > 0) || (ret == MBEDTLS_ERR_SSL_TIMEOUT)) {
+                if ((mbedtls_ms_time() - start) < 1000) {
+                    continue;
+                }
+            }
+
+            if (ret == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+                mbedtls_printf(" done, received client close notification.\n");
+            } else {
+                /* ret = 0, silent transport EOF or ret < 0 but
```
Comment (mpg) [inline programs/ssl/ssl_server2.c]: Typo (or grammar?): "but not", "and not", "except".

(Note: "but" only means "except" after certain words, like "all" etc. https://dictionary.cambridge.org/grammar/british-grammar/but)

File: programs/ssl/ssl_server2.c (hunk 2 above)
Comment (ronald-cron-arm) [reply programs/ssl/ssl_server2.c:4048]: I think that if the stdout is redirected to a file, having the "\n" does not trigger necessarily a flush.

File: tests/ssl-opt.sh (hunk 4 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh]: > From context, I'm assuming this makes the ClientHello fragmented, but otherwise I'm not sure it would be clear to me.

Yes, I am currently working on improving the documentation and the checks of which handshake messages are fragmented.

 
> Related: what's the lowest possible value we can pass to `--mtu` here? The lowest the value, the more obvious it is that the ClientHello is fragmented.

With the GnuTLS client experimentally it is 53 bytes. Below it does not work. I've settled with 128, not too small and the ClientHello should be fragmented in most (hopefully all) library configurations.

File: tests/ssl-opt.sh (hunk 5 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh]: Yes it is enough for the ClientHello to be fragmented but rather fragile. If we change the OpenSSL version we are not sure this will still work. With OpenSSL, 256 is the minimum for the MTU, we cannot go below. Thus here I am going to rather do a client authentication and check that the Certificate message is fragmented.

File: tests/ssl-opt.sh (hunk 6 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:12241]: It was inherited from the equivalent openssl server tests (DTLS proxy: 3d, openssl server). I did not find any explanation about this setting for the openssl server tests. I've removed it for the tests I am adding. Hopefully we are not going to have too much intermittent failures.

File: programs/ssl/ssl_server2.c (hunk 1 above)
Comment (ronald-cron-arm) [reply programs/ssl/ssl_server2.c]: Indeed, thanks. I've removed the call to `mbedtls_ssl_conf_read_timeout()` and adapted the documentation accordingly.

**Comment by ronald-cron-arm** [comment]
@mpg I believe I have addressed your first comments now. The PR is back to a reviewable state now.

I've reworked significantly the PR moving some commits to the preparatory PR #10575, improving the test logs and their documentation. Here is the comparison with the version you looked at and the new HEAD: https://github.com/Mbed-TLS/mbedtls/compare/1c087cfa46397dbf86cc2ccf4119e2866e15deb4..ca4ed9ca05331477e20315521b0a5acf973ff7b7.
Only the last 11 commits constitute this PR. The only changes on the C code are to improve the fragmentation/defragmentation logs and are done in #10575, not in this PR.

File: tests/ssl-opt.sh (hunk 4 above)
Comment (mpg) [reply tests/ssl-opt.sh]: Thanks for the comment!

However, perhaps I'm being too picky here, but if GnuTLS allows it, why not have a test with an MTU lower than 69, so that we know for sure the ClientHello is fragmented, instead of "likely to be fragmented in most configurations"? I'd enjoy the feeling of certainty :)

File: tests/ssl-opt.sh [hunk 8]
```diff
@@ -9957,13 +9963,59 @@ run_test    "DTLS reassembly: fragmentation, nbio, renego (gnutls server)" \
...
             -c "=> renegotiate" \
             -C "mbedtls_ssl_handshake returned" \
             -C "error" \
             -s "Extra-header:"
 
+requires_gnutls
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+run_test    "DTLS reassembly: no fragmentation (gnutls client)" \
+            "$P_SRV debug_level=2 dtls=1" \
+            "$G_NEXT_CLI -u --mtu 2048 --insecure 127.0.0.1" \
```
Comment (mpg) [inline tests/ssl-opt.sh:9978]: General note, not specifically on this line: test tend to only use powers of two for the MTU. I understand how that comes naturally, but OTOH those values could be somehow "nicer" on the code, so I think it would be good to have a few tests MTU values that are not powers of two.

(IIRC when working on testing TLS defragmentation, Minos did use a variety or values that were not always powers of two, and I think so should we, here.)

File: tests/ssl-opt.sh [hunk 9]
```diff
@@ -9957,13 +9963,59 @@ run_test    "DTLS reassembly: fragmentation, nbio, renego (gnutls server)" \
...
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+run_test    "DTLS reassembly: more fragmentation, nbio (gnutls client)" \
+            "$P_SRV debug_level=2 dtls=1 nbio=2" \
+            "$G_NEXT_CLI -u --mtu 128 --insecure 127.0.0.1" \
+            0 \
+            -s "ClientHello handshake message has been buffered and reassembled" \
+            -S "error"
+
+# No fragmentation and renegotiation tests with GnuTLS client as the feature
+# does not work properly.
```
Comment (mpg) [inline tests/ssl-opt.sh:10018]: Please tell me more about "does not work properly" :) Is it a polite way of suggesting there's  a bug on their side? More specifically, what reasons to we have to think this is not a sign there's a bug on our side?

File: tests/ssl-opt.sh [hunk 10]
```diff
@@ -9972,29 +10024,58 @@ run_test    "DTLS reassembly: no fragmentation (openssl server)" \
             -C "found fragmented DTLS handshake message" \
             -C "error"
 
+# Minimum possible MTU for OpenSSL server: 256 bytes.
+# We expect the server Certificate handshake to be fragmented and verify that
+# this is the case. Depending on the configuration, other handshake messages may
+# also be fragmented.
```
Comment (mpg) [inline tests/ssl-opt.sh:10031]: Out of curiosity, what's the size of a ClientHello in the full config? If we know it's higher than 256, I think that would be a nice thing to add to the comment.

File: tests/ssl-opt.sh [hunk 11]
```diff
@@ -9972,29 +10024,58 @@ run_test    "DTLS reassembly: no fragmentation (openssl server)" \
...
             0 \
-            -c "found fragmented DTLS handshake message" \
-            -C "error"
+            -S "found fragmented DTLS handshake message" \
+            -S "error"
+
+# Minimum possible MTU for OpenSSL server: 256 bytes.
+# We expect the server Certificate handshake to be fragmented and verify that
+# this is the case. Depending on the configuration, other handshake messages may
+# also be fragmented like the ClientHello, ClientKeyExchange or
```
Comment (mpg) [inline tests/ssl-opt.sh]: Sorry, depending on what configuration? I assume the size of the OpenSSL ClientHello does not depend on our configuration.

Is this a typo / hasty copy-paste or am I missing something?

File: tests/ssl-opt.sh [hunk 12]
```diff
@@ -10279,6 +10360,37 @@ run_test    "DTLS fragmenting: both (MTU=512)" \
...
+             key_file=$DATA_FILES_PATH/server8.key \
+             request_size=8 \
+             hs_timeout=2500-60000 \
+             mtu=128" \
+            0 \
+            -s "found fragmented DTLS handshake message" \
+            -s "fragmenting Certificate handshake message" \
+            -s "fragmenting ServerKeyExchange handshake message" \
+            -c "found fragmented DTLS handshake message" \
+            -c "fragmenting ClientHello handshake message" \
```
Comment (mpg) [inline tests/ssl-opt.sh:10391]: Nice, thanks for adding those detailed debug logs!

File: tests/ssl-opt.sh [hunk 13]
```diff
@@ -10279,6 +10360,37 @@ run_test    "DTLS fragmenting: both (MTU=512)" \
...
             -C "error"
 
+# Depending on the ciphersuite selected to encrypt the application data, the
+# maximum application data payload per record may be small with an MTU of 128.
+# For example, with TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA384, this maximum is
+# 35 bytes. We therefore reduce the size of the client request and the server
+# response in this test.
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
+requires_max_content_len 2048
+run_test    "DTLS fragmenting: both (MTU=128)" \
```
Comment (mpg) [inline tests/ssl-opt.sh:10372]: Coming back to my previous suggestions of using non-powers of two for the mtu, perhaps that would be a nice place to also try 127 and 129, since it exercises both our ability to fragment and to defragment.

File: tests/ssl-opt.sh [hunk 14]
```diff
@@ -10753,15 +10916,37 @@ requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
...
+# Depending on the ciphersuite selected to encrypt the application data, the
+# maximum application data payload per record may be small with an MTU of 128.
+# For example, with TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA384, this maximum is
+# 35 bytes. We therefore reduce the size of the client request in this test.
+requires_openssl_next
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
+requires_config_enabled PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_BASIC
+client_needs_more_time 4
+requires_max_content_len 2048
+run_test    "DTLS fragmenting: 3d, MTU=128, openssl server, DTLS 1.2" \
```
Comment (mpg) [inline tests/ssl-opt.sh]: Nice one :)

File: tests/ssl-opt.sh [hunk 15]
```diff
@@ -11904,10 +12130,49 @@ requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
...
+run_test    "DTLS proxy: 3d, gnutls client" \
+            -p "$P_PXY drop=5 delay=5 duplicate=5" \
+            "$P_SRV dgram_packing=0 dtls=1" \
+            "$G_NEXT_CLI -u --mtu 2048 --insecure 127.0.0.1" \
+            0 \
+            -s "HTTP/1.0 200 OK"
+
+# Set the MTU to 128 bytes. The ClientHello is not surely fragmented but very
+# likely. Do not set it to 56 bytes where we would be sure that the ClientHello
+# is fragmented as then experimentally the handshake fails too often.
```
Comment (mpg) [inline tests/ssl-opt.sh]: Ok, yes, for tests with the proxy emulating a very unreliable connection, I agree about not going too low.

Two points though:
1. Coming back to my previous thought that the size of the GnuTLS / OpenSSL ClientHello does not depends on our build-time config, and that we pretty much know how GnuTLS and OpenSSL have been compiled, in practice we know the size of their ClientHellos for a given version with a pretty good level of certainty.
2. I noticed another test using doing `s/5/8/` in `drop=5 delay=5 duplicate=5` and that could be an option here too if needed.

**Review by mpg — COMMENTED** [review]
I have reviewed the tests for starters, and while I left a number of comments because I'm picky, overall I'm pretty happy with the level of testing we have here.

So, I'll start looking at the code for real in my next pass.

File: tests/ssl-opt.sh (hunk 10 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:10031]: 212 bytes just for the handshake message. Thus with an MTU of 256 it is not fragmented. Otherwise this test is about the reassembly on client side of handshake messages fragmented on server side thus the ClientHello size is not really relevant here.

File: tests/ssl-opt.sh (hunk 4 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh]: We check that the server logs contain "ClientHello handshake message has been buffered and reassembled" thus if the test pass we are sure that the ClientHello was fragmented. If we change the GnuTLS version then we are not sure that the ClientHello will be fragmented but it is very likely and we will notice if this is not the case as the test will fail.

File: tests/ssl-opt.sh (hunk 11 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh]: I am just confusing myself. Fixing.

File: tests/ssl-opt.sh (hunk 15 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh]: 1. yes I've updated the comment along these lines
2. you mean that with `drop=8 delay=8 duplicate=8` the test is more likely to succeed than with `drop=5 delay=5 duplicate=5`?

Comment (mpg) [reply tests/ssl-opt.sh]: > you mean that with `drop=8 delay=8 duplicate=8` the test is more likely to succeed than with `drop=5 delay=5 duplicate=5`

Intuitively I would think so. I've seen another small-MTU test above using  `drop=8 delay=8 duplicate=8` and I assumed if was for that reason?

File: tests/ssl-opt.sh (hunk 9 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:10018]: With
```
run_test    "DTLS reassembly: fragmentation, renego (gnutls client)" \          
            "$P_SRV debug_level=2 dtls=1 renegotiation=1 renegotiate=1" \       
            "$G_NEXT_CLI -u --insecure 127.0.0.1"
```
(no MTU setting and thus no fragmentation), the GnuTLS clients errors out with:
```
*** Non fatal error: Rehandshake was requested by the peer.                     
*** Received rehandshake request                                                
*** Fatal error: The specified session has been invalidated for some reason.    
*** Rehandshake Failed: The specified session has been invalidated for some reason.
*** Fatal error: The specified session has been invalidated for some reason.    
*** Server has terminated the connection abnormally.
```
and it does not seem that we have anything to do with that. ssl_server2 sends the hello request but does not receive any new ClientHello.

**Comment by ronald-cron-arm** [comment]
> I have reviewed the tests for starters, and while I left a number of comments because I'm picky, overall I'm pretty happy with the level of testing we have here.
> 
> So, I'll start looking at the code for real in my next pass.

Thanks for your comments so far, I believe I have addressed all of them and the CI is green.

File: library/ssl_tls.c [hunk 16]
```diff
@@ -694,8 +694,16 @@ static const char *ssl_tls13_get_hs_msg_name(int hs_msg_type)
...
         case MBEDTLS_SSL_HS_CERTIFICATE_REQUEST:
             return "CertificateRequest";
+        case MBEDTLS_SSL_HS_CERTIFICATE_VERIFY:
+            return "CertificateVerify";
+        case MBEDTLS_SSL_HS_CLIENT_KEY_EXCHANGE:
+            return "ClientKeyExchange";
+        case MBEDTLS_SSL_HS_FINISHED:
+            return "Finished";
     }
     return "Unknown";
```
Comment (minosgalanakis) [inline library/ssl_tls.c:708]: How come we are covering j[ust a subset](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L565-L579) of the handshake types?

File: library/ssl_msg.c [hunk 17]
```diff
@@ -2965,6 +2967,36 @@ int mbedtls_ssl_prepare_handshake_record(mbedtls_ssl_context *ssl)
...
+         */
+        if (ssl->in_msg[0] == MBEDTLS_SSL_HS_CLIENT_HELLO &&
+            ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER &&
+            ssl->state == MBEDTLS_SSL_CLIENT_HELLO
+#if defined(MBEDTLS_SSL_RENEGOTIATION)
+            && ssl->renego_status == MBEDTLS_SSL_INITIAL_HANDSHAKE
+#endif
+            ) {
+            ssl->handshake->in_msg_seq = recv_msg_seq;
+            ssl->handshake->out_msg_seq = recv_msg_seq;
```
Comment (minosgalanakis) [inline library/ssl_msg.c]: Does this mean that the client is driving the message sequence for the server? What would happen if the client sets `ssl->handshake->out_msg_seq = recv_msg_seq -1;`

File: library/ssl_tls12_server.c [hunk 18]
```diff
@@ -869,132 +869,27 @@ static int ssl_parse_client_hello(mbedtls_ssl_context *ssl)
...
+        MBEDTLS_SSL_DEBUG_RET(1, "mbedtls_ssl_read_record ", ret);
+        return ret;
     }
 
-    buf = ssl->in_msg;
-
-    MBEDTLS_SSL_DEBUG_BUF(4, "record contents", buf, msg_len);
-
-    ret = ssl->handshake->update_checksum(ssl, buf, msg_len);
+    ret = mbedtls_ssl_update_handshake_status(ssl);
```
Comment (minosgalanakis) [inline library/ssl_tls12_server.c:893]: Should that happen after we have validated the message?

~~~~~~
if ((ssl->in_msgtype != MBEDTLS_SSL_MSG_HANDSHAKE) ||
     (buf[0] != MBEDTLS_SSL_HS_CLIENT_HELLO)) {
      MBEDTLS_SSL_DEBUG_MSG(1, ("bad client hello message"));
      return MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE;
}
ret = mbedtls_ssl_update_handshake_status(ssl);
....
~~~~~~

File: library/ssl_tls.c (hunk 16 above)
Comment (ronald-cron-arm) [reply library/ssl_tls.c:708]: This is a debug function. We cover only what we have needed so far.

**Comment by ronald-cron-arm** [comment]
@minosgalanakis @mpg I've rebased to resolve the conflicts (straightforward rebase) and added two additional commits:
Update mbedtls_ssl_handshake() documentation
Add change log

File: library/ssl_msg.c (hunk 17 above)
Comment (ronald-cron-arm) [reply library/ssl_msg.c]: As explained in the comment above we do this just for the case of the ClientHello as we don't know how many we received before. Otherwise we keep incrementing `in_msg_seq` and `out_msg_seq` independently of what we receive from the client.

File: library/ssl_tls12_server.c (hunk 18 above)
Comment (ronald-cron-arm) [reply library/ssl_tls12_server.c:893]: That's not necessary to do it after as if the message is not valid the handshake will be aborted. But it feels indeed better to do it even at the end of the function, when the ClientHello  in fully validated, like in `ssl_parse_certificate_verify()` and generally in TLS 1.3 implementation.

Comment (ronald-cron-arm) [reply library/ssl_tls12_server.c:893]: Eventually not possible to move it at the end of the function but I have moved it a bit further. See the last commit.

File: library/ssl_tls12_server.c [hunk 19]
```diff
@@ -1233,7 +1242,11 @@ static int ssl_parse_client_hello(mbedtls_ssl_context *ssl)
 #if defined(MBEDTLS_SSL_SESSION_TICKETS)
             case MBEDTLS_TLS_EXT_SESSION_TICKET:
                 MBEDTLS_SSL_DEBUG_MSG(3, ("found session ticket extension"));
-
+                /*
+                 * If the Mbed TLS ssl_ticket.c implementation is used, the
```
Comment (minosgalanakis) [inline library/ssl_tls12_server.c:1246]: Thank you for adding this context for our future selves.

File: tests/ssl-opt.sh [hunk 20]
```diff
@@ -10265,23 +10343,106 @@ run_test    "DTLS fragmenting: both (MTU=1024)" \
...
+            -c "found fragmented DTLS handshake message" \
+            -C "error"
+
+# Depending on the ciphersuite selected to encrypt the application data, the
+# maximum application data payload per record may be small with an MTU of 128.
+# For example, with TLS-ECDHE-ECDSA-WITH-AES-256-CBC-SHA384, this maximum is
+# 35 bytes. We therefore reduce the size of the client request and the server
+# response in this test and the two following tests.
+requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
+requires_max_content_len 128
```
Comment (minosgalanakis) [inline tests/ssl-opt.sh:10371]: As [discussed](https://github.com/Mbed-TLS/mbedtls/pull/10623#pullrequestreview-3854577507) in the back-port  it would be good to have a clear intent and consistent behaviour when it comes to this tests running with ECDHE_ECDSA configurations.

File: library/ssl_tls12_server.c [hunk 21]
```diff
@@ -869,29 +869,22 @@ static int ssl_parse_client_hello(mbedtls_ssl_context *ssl)
...
-     * otherwise read it ourselves manually in order to support SSLv2
-     * ClientHello, which doesn't use the same record layer format.
-     * Otherwise in a scenario of TLS 1.3/TLS 1.2 version negotiation, the
-     * ClientHello has been already fully fetched by the TLS 1.3 code and the
-     * flag ssl->keep_current_message is raised.
+     * Fetch the expected ClientHello handshake message. Do not ask
+     * mbedtls_ssl_read_record() to update the handshake digest, to align
+     * with cases where the ClientHello may already have been fetched in
+     * ssl_tls13_process_client_hello() or as a post-handshake message
+     * (renegotiation).
```
Comment (mpg) [inline library/ssl_tls12_server.c]: I find this comment a bit confusing - more specifically, I completely misunderstood it initially. Suggestion:
```
     * Fetch the expected ClientHello handshake message. Do not ask
     * mbedtls_ssl_read_record() to update the handshake digest, because
     * in cases where the ClientHello may already have been fetched in
     * ssl_tls13_process_client_hello() or as a post-handshake message
     * (renegotiation) we'll have to do it ourselves, and it's simpler to do so
     * unconditionally than to try to remember if we need to.
```

File: library/ssl_tls12_server.c (hunk 18 above)
Comment (mpg) [reply library/ssl_tls12_server.c:893]: Counter-point: there's also an argument for keeping this close to the call to `mbedtls_ssl_read_record(..., 0)`, because this is normally part of `read_record()` and doing it manually instead is a bit of a special case.

(Note: the normal path of doing it as part of `read_record` means we normally update the checksum well before we have validated the contents of the handshake message, but as Ronald said, I think this is fine since any validation issue with result in aborting the handshake anyway.)

File: library/ssl_msg.c [hunk 22]
```diff
@@ -2967,6 +2967,27 @@ int mbedtls_ssl_prepare_handshake_record(mbedtls_ssl_context *ssl)
             return MBEDTLS_ERR_SSL_INVALID_RECORD;
         }
 
+        /*
+         * When establishing the connection, the client may go through a series
+         * of ClientHello and HelloVerifyRequest requests and responses. The
+         * server does not keep any trace of these initial round trips as
+         * intended: minimum allocated ressources as long as the reachability
```
Comment (mpg) [inline library/ssl_msg.c]: Minor: "the server does not [do X] as intended" may be interpret either as "the server doesn't to what we intended it to" or as "the server doesn't do X and we we don't intend it to". Suggesting "the server intentionally does not keep track" to make it less ambiguous.

File: library/ssl_tls12_server.c [hunk 23]
```diff
@@ -950,34 +950,6 @@ static int ssl_parse_client_hello(mbedtls_ssl_context *ssl)
...
-    if (ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM) {
-        /*
-         * Copy the client's handshake message_seq on initial handshakes,
-         * check sequence number on renego.
-         */
-#if defined(MBEDTLS_SSL_RENEGOTIATION)
-        if (ssl->renego_status == MBEDTLS_SSL_RENEGOTIATION_IN_PROGRESS) {
-            /* This couldn't be done in ssl_prepare_handshake_record() */
-            unsigned int cli_msg_seq = (unsigned int) MBEDTLS_GET_UINT16_BE(ssl->in_msg, 4);
-            if (cli_msg_seq != ssl->handshake->in_msg_seq) {
```
Comment (mpg) [inline library/ssl_tls12_server.c:1028]: Note to self: we have a [similar check](https://github.com/Mbed-TLS/mbedtls/blob/development/library/ssl_msg.c#L4068) in `mbedtls_ssl_prepare_handshake_record()` which is called by `mbedtls_ssl_handle_message_type()` itself called by `mbedtls_ssl_read_record()`.

File: library/ssl_tls12_server.c (hunk 18 above)
Comment (ronald-cron-arm) [reply library/ssl_tls12_server.c:893]: I've moved it back and adapted the associated comment.

File: library/ssl_tls12_server.c (hunk 23 above)
Comment (mpg) [reply library/ssl_tls12_server.c:1028]: Oops, wrong link, meant https://github.com/Mbed-TLS/mbedtls/blob/development/library/ssl_msg.c#L2972

Comment (mpg) [reply library/ssl_tls12_server.c:1028]: Note: as discussed elsewhere, for the record, I think I was wrong and the check is not done in `prepare_handshake_record()` for the ClientHello that initiates a renego, because at this point `is_handshake_over()` will still return true.
